ToolBox/PsychSignalDataConverter/PsychSignalDataConverter.cs

```diff
@@ -0,0 +1,448 @@
+        /// <param name="endDateUtc">Date to stop parsing files</param>
+        /// <remarks>Inclusive on the lower bound, and exclusive on the upper bound</remarks>
+        public void ConvertFrom(DateTime startDateUtc, DateTime endDateUtc)
+        {
+            if (_disposedValue)
+            {
+                throw new ObjectDisposedException("PsychSignalDataConverter has already been disposed");
+            }
+
+            // Filter for files by name and bounds, then order by date
+            var files = _rawSourceDirectory.GetFiles("*.csv", SearchOption.TopDirectoryOnly)
+                .Where(
+                    x =>
+                    {
+                        DateTime fileDate;
+                        if (!DateTime.TryParseExact(x.Name.Substring(0, 11), "yyyyMMdd_HH", CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal, out fileDate))
+                        {
+                            return false;
+                        }
+
+                        return fileDate >= startDateUtc && fileDate < endDateUtc;
+                    }
+                )
+                .OrderBy(x => DateTime.ParseExact(x.Name.Substring(0, 11), "yyyyMMdd_HH", CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal))
+                .ToList();
+
+            var fileCount = files.Count();
+            var i = 0;
+
+            foreach (var rawFile in files)
+            {
+                i++;
+                Log.Trace($"PsychSignalDataConverter.ConvertFrom(): Reading file {rawFile.Name} (file {i}/{fileCount})");
+                Log.Trace($"PsychSignalDataConverter.ConvertFrom(): Begin converting {rawFile.Name}");
+
+                using (var stream = rawFile.OpenRead())
+                {
+                    Convert(stream);
+                }
+
+                Log.Trace($"PsychSignalDataConverter.Convert(): Finished converting {rawFile.Name}");
+            }
+
+            Dispose();
+        }
+
+        /// <summary>
+        /// Converts a single day's data. We start processing data
+        /// from the 0th hour of the day and finish at the 23rd hour of the day
+        /// </summary>
+        /// <param name="date">Date to convert files for</param>
+        public void ConvertDate(DateTime date)
+        {
+            ConvertFrom(date.Date, date.Date.AddDays(1));
+        }
+
+        /// <summary>
+        /// Converts gzipped backfill data. Reads all *.gz files in the raw directory and attempts to convert them
+        /// </summary>
+        public void ConvertHistoricalData()
+        {
+            foreach (var archive in _rawSourceDirectory.GetFiles("*.gz", SearchOption.TopDirectoryOnly))
+            {
+                Log.Trace($"PsychSignalDataConverter.ConvertBackfill(): Begin converting historical data for file: {archive.FullName}");
+                using (var archiveStream = StreamProvider.ForExtension(".gz").Open(archive.FullName).First())
```

Comment:
nit - if we're assuming there's _exactly_ one of something, we should be using `Single()` instead of `First()` so that if our assumption is wrong, we'll get an error.